src/community/geopkg/src/main/java/org/geoserver/geopkg/wps/OWSContextWriter.java

```diff
@@ -0,0 +1,433 @@
+        if (layerStyles.size() < 2) {
+            return;
+        }
+
+        // get all the layer groups, find the ones containing the layers and styles
+        // that we have dumped
+        List<LayerGroupInfo> layerGroups = getGroupsByLayerStyles();
+        for (LayerGroupInfo layerGroup : layerGroups) {
+            writeStyleContext(layerGroup, layerStyles, contents.getName());
+        }
+    }
+
+    private void writeStyleContext(
+            LayerGroupInfo layerGroup, Map<String, LayerInfo> layerStyles, String packageName)
+            throws IOException {
+        // setup the template model
+        HashMap<String, Object> model = new HashMap<>();
+        model.put("groupId", layerGroup.prefixedName());
+        String groupTitle =
+                Optional.ofNullable(layerGroup.getTitle()).orElse(layerGroup.prefixedName());
+        model.put("groupTitle", groupTitle);
+        model.put("now", getCurrentISOTimestamp());
+        model.put("contact", gs.getGlobal().getSettings().getContact());
+        model.put("layers", getLayersModel(layerGroup, layerStyles));
+        model.put("packageName", packageName);
+
+        String stylesheetContext = process(STYLESHEETS_TEMPLATE, model);
+
+        try {
+            GeoPkgMetadata metadata =
+                    new GeoPkgMetadata(
+                            Scope.Undefined,
+                            OWS_CONTEXT_JSON_URI,
+                            GEO_JSON_MIME,
+                            stylesheetContext);
+            metadatas.addMetadata(metadata);
+
+            GeoPkgMetadataReference reference =
+                    new GeoPkgMetadataReference(
+                            GeoPkgMetadataReference.Scope.GeoPackage,
+                            null,
+                            null,
+                            null,
+                            new Date(),
+                            metadata,
+                            null);
+            metadatas.addReference(reference);
+
+            GeoPkgSemanticAnnotation annotation =
+                    new GeoPkgSemanticAnnotation(
+                            STYLESHEET_SA_TYPE,
+                            "OGC OWS Context GeoJSON for " + groupTitle,
+                            STYLESHEET_SA_URI);
+            annotations.addAnnotation(annotation);
+
+            GeoPkgAnnotationReference ar =
+                    new GeoPkgAnnotationReference(
+                            "gpkg_metadata", "id", metadata.getId(), annotation);
+            annotations.addReference(ar);
+        } catch (SQLException e) {
+            throw new IOException(e);
+        }
+    }
+
+    private List<Map<String, Object>> getLayersModel(
+            LayerGroupInfo layerGroup, Map<String, LayerInfo> layerStyles) throws IOException {
+        List<Map<String, Object>> result = new ArrayList<>();
+        List<PublishedInfo> layers = layerGroup.getLayers();
```

Comment:
thought: Are style drive groups going to cause confusion here? Thinking of an style that draws the same features multiple times to establish a z-order.

Reply:
This is exactly what the OWS context in question needs to do, dump the right drawing order and associate it to styles. So a  dump of the layer groups structure is exactly that is needed. However, testing we found a small issue, in GeoServer the group is defined in painter's order, while OWS context unfortunately defined the opposite order, the first listed layer must be painted on top. Will have to make a small change for it.

Reply:
Order has been fixed now.